### PR TITLE
rpc: improve `getreceivedby{address,label}` performance

### LINKED ISSUE #23645: getreceivedby[address, label] rpc performance proposal
Hi,

**Problem**

The refactoring done in #17579 introduced some performance degradation for the rpc commands `getreceivedbyaddress` and `getreceivedbylabel`, which prevented an upgrade to versions >=0.21.0 for us without custom patches that revert this PR. The requests take notably longer to complete. This is especially painful when using big wallet files or issuing a large number of requests in a short period of time.
I know that the foremost goal of Bitcoin Core may not be performance, but given that these rpc commands are some of the most prominent ones needed when interacting with wallets via rpc, I propose to do some changes here.

**Analysis**

Currently, a set of search address is generated from the scriptpubkeys and then these addresses are compared against all addresses in the wallet, which means for each transaction in the wallet, the time consuming conversion from scriptpubkey to address needs to happen (in `ExtractDestination(...)`).

**Solution Proposal**

Given that the majority of use cases comprises searching for a small set of addresses in a large set of addresses (the wallet), it makes more sense to rather create a set of scriptpubkeys from the search addresses and then compare each scriptpubkey in the wallet against this set. This is similar to how it was done before #17579, except that we now use a set, which means a search for multiple addresses (corresponding to a single label) stays possible and the refactoring can be more or less kept as is.

I may not have time soon to implement this properly, but what do you think (conceptually)?

### PR DESCRIPTION
**Pull request by theStack**
The RPC calls `getreceivedbyaddress`/`getreceivedbylabel` both use the internal helper function `GetReceived` which was introduced in PR #17579 to deduplicate tallying code. For every wallet-related transaction output, the following unnecessary operations are currently performed in the tally loop, leading to a quite bad performance (as reported in #23645):
- converting from CScript -> TxDestination (`ExtractDestination(...)`), converting from TxDestination -> CScript (`CWallet::IsMine(const CTxDestination& dest)`); this can be avoided by directly using output scripts in the search set instead of addresses (first commit)
- checking if the iterated output script belongs to the wallet by calling `IsMine`; this can be avoided by only adding addresses to the search set which fulfil `IsMine` in the first place (second commit)

### Benchmark results
The functional test [wallet_pr23662.py](https://github.com/theStack/bitcoin/blob/pr23662_benchmarks/test/functional/wallet_pr23662.py) (not part of this PR) creates transactions with 15000 different addresses:
- 5000 outputs (500 txs with 10 outputs each) with label set, IsMine set (received)
- 5000 outputs (500 txs with 10 outputs each) with label set, IsMine not set (sent)
- 5000 outputs (500 txs with 10 outputs each) without label set, IsMine not set (sent)

Then, the time is measured for calling `getreceivedbyaddress` and `getreceivedbylabel`, the latter with two variants. Results on my machine:

| branch             | `getreceivedbyaddress` (single) | `getreceivedbylabel` (single) | `getreceivedbylabel` (10000) |
|--------------------|---------------------------------|-------------------------------|------------------------------|
| master             |             406.13ms            |          425.33ms             |          446.58ms            |
| PR (first commit)  |             367.18ms            |          365.81ms             |          426.33ms            |
| PR (second commit) |               3.96ms            |            4.83ms             |          339.69ms            |

Fixes #23645.

### CODE REVIEW

**Comment by kristapsk** [comment]
> if users put a huge number of addresses under the same label. Not sure how that is used in production, I always assumed that there is a one-by-one label->address relation and having more than one address referred to by a label doesn't make too much sense.

JoinMarket users labels to mark addresses in a Bitcoin Core watchonly wallet as belonging to a specific JoinMarket wallet. All addresses of the same JoinMarket wallet will have the same label in Bitcoin Core watchonly wallet. It used accounts before, until they were deprecated.

**Comment by beegater** [comment]
ACK!

Great, thanks a lot for addressing this so quickly!

**Review by alirezaayande — APPROVED** [review]

File: src/wallet/rpcwallet.cpp [hunk 1]
```diff
@@ -504,12 +504,17 @@ static RPCHelpMan listaddressgroupings()
...
 static CAmount GetReceived(const CWallet& wallet, const UniValue& params, bool by_label) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet)
 {
-    std::set<CTxDestination> address_set;
+    std::set<CScript> output_scripts;
 
     if (by_label) {
         // Get the set of addresses assigned to label
         std::string label = LabelFromValue(params[0]);
-        address_set = wallet.GetLabelAddresses(label);
+        for (auto& address : wallet.GetLabelAddresses(label)) {
```
Comment (MarcoFalke) [inline src/wallet/rpcwallet.cpp:512]: why is this a mutable reference?

**Review by maflcko — APPROVED** [review]
lgtm

File: src/wallet/rpcwallet.cpp (hunk 1 above)
Comment (theStack) [reply src/wallet/rpcwallet.cpp]: Thanks, good catch, fixed.

**Review by shaavan — APPROVED** [review]
crACK 7589305280a7b0f43a4dd500b9fc72e070104bd2

Thanks for addressing it. These updates make the code more clear to understand and straightforward to run.

> as it could actually hurt performance if users put a huge number of addresses under the same label.

Though I think this refactoring would be very advantageous for an average user, this statement still worries me a little. Maybe if someone might have some insights into how frequent a user is to put multiple addresses under the same label, they could share them.

**Comment by theStack** [comment]
Rebased on master (necessary as the function modified in this PR has moved from `./src/wallet/rpcwallet.cpp` to `./src/wallet/rpc/coins.cpp`).

@shaavan: Thanks for your review! The performance worries on the second commit concerns transactions that are in `wallet.mapWallet` but don't satisfy the `IsMine` predicate (which should be all txs sent from the wallet to an address not owned by the wallet) ; if the search in the output script set (`output_scripts.count(...)`) takes longer than the `IsMine` check, then on master the loop iteration would be faster for those transactions. On the other hand a search in std::set takes O(log n), so it shouldn't be that bad.

Would be very interested in further opinions.

**Review by amadeuszpawlik — APPROVED** [review]
Concept & approach ACK

**Review by stratospher — COMMENTED** [review]
Concept ACK.
 - [x]  a7b65af
      - Output scripts are used instead of addresses to perform the operations.
      - For the 'getreceivedbylabel' case, all the addresses with the given label are iterated through to insert all the `output_scripts`. For the 'getreceivedbyaddress' case, the required address's output script is inserted into `output_scripts`.
      - When tallying all the transactions, `output scripts` can be directly used to compute the total amount at an address/label in the wallet.

  - [x]  f336ff7
       - For the 'getreceivedbylabel' case, when iterating through all the addresses with the given label to insert the `output_scripts`, only `output_scripts` that belong to the wallet are added.
       - When tallying all the transactions, `output scripts` are used to compute the total amount at an address/label(and it's already known that they belong to the wallet).

**Comment by achow101** [comment]
I don't think it's even necessary to use `IsMine`. We already filter for addresses that are ours within the address book itself by setting `CAddressBookData.purpose` to `receive`. So you can just check for that string rather than doing `IsMine`. It will certainly be more efficient.

**Comment by theStack** [comment]
> I don't think it's even necessary to use `IsMine`. We already filter for addresses that are ours within the address book itself by setting `CAddressBookData.purpose` to `receive`. So you can just check for that string rather than doing `IsMine`. It will certainly be more efficient.

Thanks for the hint! In the current version of the PR though, `IsMine` is not called in the hot spot (mapWallet/CTxOut iteration loop) anymore. It may still make sense to check for the `CAddressBookData.purpose` instead of using `IsMine` in the preparation loop, which creates the "needles" of the search. Will take a deeper look on the weekend.

**Comment by beegater** [comment]
Can this please be merged? It would help a lot!

**Comment by kristapsk** [comment]
I'm worried about "the second commit is potentially more controversial, as it could actually hurt performance if users put a huge number of addresses under the same label", as this is how JoinMarket uses Core wallets. Guess some benchmarking is needed here, how much is performance gains / losses in different scenarios.

**Comment by beegater** [comment]
> I'm worried about "the second commit is potentially more controversial, as it could actually hurt performance if users put a huge number of addresses under the same label", as this is how JoinMarket uses Core wallets. Guess some benchmarking is needed here, how much is performance gains / losses in different scenarios.

For the second commit this may be true. But the version that has now been on master for several years is FAR slower under any scenario than the solution in this initial commit. And just pushing the merge forward until some benchmarking is set up or not will just mean the slow solution will stay in all upcoming versions.
Since nobody else has risen awareness to this bug since all that time it is on master, I'd say that the 10x degraded performance of this rpc was not relevant to anyone anyway, so merging this PR which provides a speed up for most scenarios should be fine.

**Comment by theStack** [comment]
Finally did some benchmarking. The functional test [wallet_pr23662.py](https://github.com/theStack/bitcoin/blob/pr23662_benchmarks/test/functional/wallet_pr23662.py) (not part of this PR) creates transactions with 15000 different addresses:
- 5000 outputs (500 txs with 10 outputs each) with label set, IsMine set (received)
- 5000 outputs (500 txs with 10 outputs each) with label set, IsMine not set (sent)
- 5000 outputs (500 txs with 10 outputs each) without label set, IsMine not set (sent)

Then, the time is measured for calling `getreceivedbyaddress` and `getreceivedbylabel`, the latter with two variants. Results on my machine:

| branch | `getreceivedbyaddress` (single) | `getreceivedbylabel` (single) | `getreceivedbylabel` (10000) |
|---------|-------------------------------|---------------------------------------|----------------------------|
| master | 406.13ms| 425.33ms | 446.58ms |
| PR (first commit) | 367.18ms| 365.81ms| 426.33ms |
| PR (second commit) | 3.96ms | 4.83ms | 339.69ms |

Seems like my worries ("the second commit is potentially more controversial, as it could actually hurt performance if users put a huge number of addresses under the same label") are not justified and most of the performance gain is indeed achieved by the second commit. Improvement ideas to the benchmark are very welcome, though I'd consider this pretty safe to merge at this point.

**Comment by achow101** [comment]
ACK f336ff7f213564909cf5f9742618cc6ec87600fd

**Review by w0xlt — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/23662/commits/f336ff7f213564909cf5f9742618cc6ec87600fd

This change makes the code clearer and better performing.

**Comment by kristapsk** [comment]
Comments by @theStack and @achow101 calmed me down, but will try to benchmark myself just in case too.

**Review by furszy — APPROVED** [review]
Code ACK f336ff7f

Plus, probably something like https://github.com/furszy/bitcoin-core/commit/e8d860742ca1c834c0e4a31a38bd14ab3d681f5d would be good to include here or on a follow-up PR.

Basically, inside `GetReceived`, we can return early if the label provided by `getreceivedbylabel` is not in the address book. Otherwise, we walk through all the wallet txs + outputs for no reason (`output_scripts` is empty).

**Comment by theStack** [comment]
@furszy: Thanks for reviewing!

> Plus, probably something like [furszy@e8d8607](https://github.com/furszy/bitcoin-core/commit/e8d860742ca1c834c0e4a31a38bd14ab3d681f5d) would be good to include here or on a follow-up PR.
> 
> Basically, inside `GetReceived`, we can return early if the label provided by `getreceivedbylabel` is not in the address book. Otherwise, we walk through all the wallet txs + outputs for no reason (`output_scripts` is empty).

This seems to be an optimization for a very special case (passing a label that _is_ in the address book, but doesn't apply to any address that is ours), hence I think this would fit better for a follow-up PR. Would you mind opening one? If you want, can already do that now, branching off of this PR.

**Comment by furszy** [comment]
Hey, sure. Will do it.

> This seems to be an optimization for a very special case (passing a label that is in the address book, but doesn't apply to any address that is ours)

Actually, can happen when the user inputs a label that is not in the address book as well (aka `wallet.GetLabelAddresses(label)` returns an empty set).

**Comment by theStack** [comment]
> > This seems to be an optimization for a very special case (passing a label that is in the address book, but doesn't apply to any address that is ours)
> 
> Actually, can happen when the user inputs a label that is not in the address book as well (eg: `wallet.GetLabelAddresses(label)` returning an empty set).

Good point. So it would be an improvement of the user-interface, which is maybe even more important; right now if the passed label doesn't exist (due to e.g. a typo from the user), the amount of 0 is returned instead of an error.

**Comment by furszy** [comment]
yeah, for that I added an error there "Address/es not found in wallet" (a general error for both `getreceivedbylabel` and `getreceivedbyaddress` if the address/label is not in the wallet).

But we could customize it a bit more too; have an error string targeting `getreceivedbyaddress` -> "<address> not found in wallet" and another one for `getreceivedbylabel` -> "<label> not found in wallet".

In any case, just opened #25122. Will rebase it on master once this one gets merged.

**Pull request merged**